### PR TITLE
ddl: add a channel to limit multiple DDL jobs writing at the same time

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve? 
When performing DDL operations on multiple connections at the same time, you may encounter a "Write conflict" error.

### What is changed and how it works?
Adding a channel on a single TiDB to limit multiple DDL jobs writing to the DDL job list at the same time results in "Write conflict".

### Check List 

Tests 

 - Benchmark test
```
func addCreateTableJob(ctx sessionctx.Context, d *ddl, dbInfo *model.DBInfo, tblInfo *model.TableInfo) *model.Job {
	job := &model.Job{...}
	task := &limitJobTask{job, make(chan error)}
	d.limitJobCh <- task
	err := <-task.err
...
	return job
}

func BenchmarkAddDDLs(b *testing.B) {
...
	taskCnt := 1
	count := 200
	tblInfos := make([]*model.TableInfo, 0, count*taskCnt)
	for i := 0; i < count*taskCnt; i++ {
		tblInfo := testTableInfo(d, fmt.Sprintf("t_%d", i), 2)
		tblInfos = append(tblInfos, tblInfo)
	}
	wg := sync.WaitGroup{}
	runDDLsFunc := func(start, end int) {
		defer wg.Done()
		for i := start; i < end; i++ {
			addCreateTableJob(ctx, d, dbInfo, tblInfos[i])
		}
	}
	
	b.ResetTimer()
	wg.Add(taskCnt)
	for i := 0; i < taskCnt; i++ {
		go runDDLsFunc(i*count, (i+1)*count)
	}
	wg.Wait()
	b.StopTimer()
}
```

WC is the error of `write confilct`.

| taskCnt | count | spend time(after) | WC count(after) | spend time(before) | WC count(before) |
| :-: | :-: | :-: | :-: | :-: | :-: |
| 1 | 200 | 0.117ns | 7 | 0115ns | 6 |
| 3 | 200 | 0.254ns | 18 | 1.092 | 147 |
| 5 | 200 | 0.419s | 41 | 10.387 | 1210 |
| 5 | 400 | 1.894s | 148 | 68.745 | 3509 | 

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Adding a channel on a single TiDB to limit multiple DDL jobs writing to the DDL job list at the same time results in "Write conflict".

### CODE REVIEW

File: ddl/ddl_test.go [hunk 1]
```diff
@@ -70,6 +71,13 @@ func (d *ddl) generalWorker() *worker {
 // It only starts the original workers.
 func (d *ddl) restartWorkers(ctx context.Context) {
 	d.quitCh = make(chan struct{})
+	go util.WithRecovery(
+		func() { d.limitDDLJobs() },
+		func(r interface{}) {
+			logutil.BgLogger().Error("[ddl] DDL add batch DDL jobs meet paninc",
```
Comment (bb7133) [inline ddl/ddl_test.go]:
```suggestion
			logutil.BgLogger().Error("[ddl] DDL add batch DDL jobs meet panic",
```

**Comment by bb7133** [comment]
Please explain the meaning of ”taskCnt" and "Count" in the table in your PR Description.

**Comment by zimulala** [comment]
@bb7133 ”taskCnt" and "Count" are the variables in the benchmark.

File: ddl/ddl_test.go (hunk 1 above)
Comment (zimulala) [reply ddl/ddl_test.go]: Done.

**Comment by zimulala** [comment]
PTAL @bb7133 @crazycs520 @tiancaiamao

File: ddl/ddl_worker.go [hunk 2]
```diff
@@ -185,28 +185,62 @@ func buildJobDependence(t *meta.Meta, curJob *model.Job) error {
...
+			job := task.job
+			job.Version = currentVersion
+			job.StartTS = txn.StartTS()
+			job.ID = ids[i]
+			if err = buildJobDependence(t, job); err != nil {
+				return errors.Trace(err)
+			}
 
-		return errors.Trace(err)
+			if job.Type.String() == model.AddIndexStr || job.Type.String() == model.AddPrimaryKeyStr {
```
Comment (crazycs520) [inline ddl/ddl_worker.go]: `if job.Type == model.ActionAddIndex || job.Type == model.ActionAddPrimaryKey {`

File: ddl/ddl_worker.go [hunk 3]
```diff
@@ -185,28 +185,62 @@ func buildJobDependence(t *meta.Meta, curJob *model.Job) error {
 	return nil
 }
 
-// addDDLJob gets a global job ID and puts the DDL job in the DDL queue.
-func (d *ddl) addDDLJob(ctx sessionctx.Context, job *model.Job) error {
+func (d *ddl) limitDDLJobs() {
+	for {
+		tasks := make([]*limitJobTask, 0, batchAddingJobs)
+		select {
+		case task := <-d.limitJobCh:
```
Comment (tiancaiamao) [inline ddl/ddl_worker.go]: I do not quite understand how this limits the task rate.
By let multiple worker conflict on the same channel and increase the runtime lock spin?

File: ddl/ddl_worker.go [hunk 4]
```diff
@@ -185,28 +185,62 @@ func buildJobDependence(t *meta.Meta, curJob *model.Job) error {
...
 
-		return errors.Trace(err)
+			if job.Type == model.ActionAddIndex || job.Type == model.ActionAddPrimaryKey {
+				jobKey := meta.AddIndexJobListKey
+				err = t.EnQueueDDLJob(job, jobKey)
+			} else {
+				err = t.EnQueueDDLJob(job)
+			}
+			if err != nil {
+				return errors.Trace(err)
```
[the remark this answers is not in the document]
Comment (zimulala) [reply ddl/ddl_worker.go]: Yes, but it returns in `RunInNewTxn`. We will do `task.err <- err` in line238.

File: ddl/ddl_worker.go (hunk 3 above)
Comment (zimulala) [reply ddl/ddl_worker.go]: Write conflicts occur when multiple DDL jobs are written to the DDL job list at the same time.
We use this channel to reduce such conflicts.

File: ddl/ddl_worker.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply ddl/ddl_worker.go]: Oh, I see

**Comment by tiancaiamao** [comment]
LGTM

File: ddl/ddl_worker.go [hunk 5]
```diff
@@ -185,28 +185,62 @@ func buildJobDependence(t *meta.Meta, curJob *model.Job) error {
 	return nil
 }
 
-// addDDLJob gets a global job ID and puts the DDL job in the DDL queue.
-func (d *ddl) addDDLJob(ctx sessionctx.Context, job *model.Job) error {
+func (d *ddl) limitDDLJobs() {
+	for {
+		tasks := make([]*limitJobTask, 0, batchAddingJobs)
```
Comment (djshow832) [inline ddl/ddl_worker.go]: This line should be moved out of the loop.

File: ddl/ddl_worker.go [hunk 6]
```diff
@@ -185,28 +185,62 @@ func buildJobDependence(t *meta.Meta, curJob *model.Job) error {
...
-			return errors.Trace(err)
-		}
-		err = t.EnQueueDDLJob(job)
+		for i, task := range tasks {
+			job := task.job
+			job.Version = currentVersion
+			job.StartTS = txn.StartTS()
+			job.ID = ids[i]
+			if err = buildJobDependence(t, job); err != nil {
+				return errors.Trace(err)
```
Comment (djshow832) [inline ddl/ddl_worker.go]: If one task fails, following tasks are discarded?

Comment (zimulala) [reply ddl/ddl_worker.go]: Yes, but this function will fail only we get jobs failed from. Therefore, if this job gets jobs failed, the other job should also fail.

File: ddl/ddl_worker.go (hunk 5 above)
Comment (zimulala) [reply ddl/ddl_worker.go]: Done.

File: ddl/ddl_worker.go (hunk 6 above)
Comment (zimulala) [reply ddl/ddl_worker.go]: PTAL @djshow832

**Comment by zimulala** [comment]
PTAL @coocood

**Comment by zimulala** [comment]
PTAL @coocood @AilinKid

File: ddl/ddl_worker.go [hunk 7]
```diff
@@ -195,28 +195,62 @@ func buildJobDependence(t *meta.Meta, curJob *model.Job) error {
...
 		}
-		job.StartTS = txn.StartTS()
-		if err = buildJobDependence(t, job); err != nil {
-			return errors.Trace(err)
-		}
-		err = t.EnQueueDDLJob(job)
+		for i, task := range tasks {
+			job := task.job
+			job.Version = currentVersion
+			job.StartTS = txn.StartTS()
```
Comment (AilinKid) [inline ddl/ddl_worker.go]: batch DDL jobs share the same txn.StartTS()? unless this ddl is in same statement.

Comment (zimulala) [reply ddl/ddl_worker.go]: Yes. But not necessarily the same statement. 
And this field has no special use in TiFilash and Tools.

Comment (AilinKid) [reply ddl/ddl_worker.go]: Seems ok

**Pull request merged**